# Combinatorial Thinking - Staircase Problem for taking 1 or 2 steps at a time

The stair-case-problem, taking 1 or 2 steps at a time, questions how many distinct ways there are to take all stairs.
The [**most frequent solution**](https://quanticdev.com/algorithms/dynamic-programming/staircase-problems/) to that problem, which also provides a solution for taking up to *m* steps, is based on the Fibonacci sequence. However, without applying a recursive or iterative cycle-evaluation a solution resulting directly to the Fibonacci sequence is hardly possible or intuitive. As mentioned, applying firstly a cylce-evaluation, which sets up all combinations reveals secondly the Fibonacci sequence. A more intuitive modelling approach can be achieved via combinatorics. In the following part a combinatorial solution is introduced, which also should help for abstraction in **Combinatorial Thinking**.

 - Part 3: Deriving a general formula for *n* stairs and taking 1 or 2 steps at a time.
 - Part 4: Comparing results with the Fibonacci-solution.
 - Part 5: Considering additionally a pause option when having number of stairs' actions of taking 1, 2 or 0 steps at a time.

![](https://res.cloudinary.com/practicaldev/image/fetch/s--s32xY0FA--/c_limit%2Cf_auto%2Cfl_progressive%2Cq_auto%2Cw_880/https://dev-to-uploads.s3.amazonaws.com/i/rdmbuufsp88bldsqf7fp.png)

# Imports

In [5]:
import sys
sys.path.insert(0, 'lib\\')
import pandas as pd 
import numpy as np
from matplotlib import pyplot as plt
import scipy.special

# Deriving a general formula for *n* stairs and taking 1 or 2 steps at a time
Part 3. General problem statement.

Let *n* be the number of stairs, greater than 0, and let a stair-taker either take 1 or 2 steps at a time. Furthermore, a field of length *n* is introduced, which can also contain one or more 0's. These 0's are interpreted as a pause which the step-taker takes. This mentioned field can then be used for an iterative solution in which simply all possible combinations are evaluated (cycle-evaluation), but this numerical approach is suboptimal due to its time complexity of O(2^n). The [**recursion with memoization**](https://quanticdev.com/algorithms/dynamic-programming/staircase-problems/) can bring the time complexity down to O(n). For the part 3 the 0's can be completely ignored.

The modelling approach is divided into three steps resp. levels:
 1. First level:  Computing the number of combinations without order (part 3.1).
 2. Second level: Computing the number of combinations with order, but without considering any 0 for a pause (part 3.2).
 3. Third level:  Computing the number of combinations with order and with considering 0 for a pause (part 5).
 
These levels build up on each other. 

For level one all possible unique combinations without considering the order are computed. Here, zeros are ignored completely and where a step-taker takes 2 steps instead of 1 is irrelevant.
The second level computes all permutations for each combination from the first level. Here, zeros are ignored completely and where a step-taker takes 2 steps instead of 1 is relevant.
Finally the third level computes for each combination from the second level all further permutations, where a stair-taker can take their pause. Here, ther order of the zeros and where a step-taker takes 2 steps instead of 1 are relevant.

## Computing the number of combinations without order for *n* stairs and taking 1 or 2 steps at a time
Part 3.1. Combinations of the **first level**. The order, where 2 steps instead of 1 are as well as zeros are irrelavant.

The unique number of combinations with given elements is the first thing which needs to be determined. Considering the order is then just permutating these elements for each combination (second level)), which was determined in the first level.

Consider 10 stairs, which implies a field of size 10. Consider the most simple case where the stair-taker simply always takes one step:

$$1.\hspace{10mm}[1,1,1,1,1,1,1,1,1,1]$$

For each time the stair-taker takes two steps instead of one also a pause(a 0) needs to be added to the field:
$$2.\hspace{10mm}[2,1,1,1,1,1,1,1,1,0]$$
$$3.\hspace{10mm}[2,2,1,1,1,1,1,1,0,0]$$
$$\hspace{10mm}...$$
$$5.\hspace{10mm}[2,2,2,2,1,1,0,0,0,0]$$
$$6.\hspace{10mm}[2,2,2,2,2,0,0,0,0,0]$$

Here it should be emparized that the order does not matter hence,

$$2.\hspace{10mm}[2,1,1,1,1,1,1,1,1,0] = [1,2,1,1,1,1,1,1,1,0] = [1,1,2,1,1,1,1,1,0] = ...$$

The 0's can be ignored completely here.

One main conclusion can be taken from this: 
 - If the maximum number of steps is 2 and in total there are 10 stairs then maximal five 2's fit in 10 stairs. If there would be 11 stairs then still only five 2's would fit completely in them hence, the number of 2's fitting in *n* stairs is therefore simply 
 
$$floor(n / 2)$$ 
 
Considering additionally the one combination with no 2 at all leads to:

$$(3.1)\hspace{10mm}NumOfCombinationsWithoutOrder(n) = floor(n / 2) + 1 $$  


In [6]:
def numCombinationsWithoutOrderAndWithoutZeros(stairs):
    return (int(stairs / 2) + 1)

In [7]:
n = 10
print('Computing the number of combinations without order and without zeros for', n, 'stairs is', numCombinationsWithoutOrderAndWithoutZeros(n))

Computing the number of combinations without order and without zeros for 10 stairs is 6


## Computing the number of combinations with order for *n* stairs and taking 1 or 2 steps at a time
Part 3.2. Combinations of the **second level**. The order, where 2 steps insted of 1 are taken is relevent, but zeros are ignored.

For each unique combination with given elements the number of permutations need to be determined hence, rearranging all possible orders for taking 1 and/or 2 steps.

Considering the second combination with one 2 resp. once take two steps:
$$\hspace{0mm}[2,1,1,1,1,1,1,1,1,0]$$ 
$$\hspace{0mm}[1,2,1,1,1,1,1,1,1,0]$$
$$\hspace{0mm}...$$
$$\hspace{0mm}[1,1,1,1,1,1,1,2,1,0]$$
$$\hspace{0mm}[1,1,1,1,1,1,1,1,2,0]$$

In total there are 9 permutations, which lead to a different ordering (ignoring the 0). So, the number of permutations for each unique combination (3.1) depends on the total number of taking 1 and 2 steps. In the previous example it was one 2 out of nine actions the step-taker took. Therefore, the combination [2,1,1,1,1,1,1,1,1,0] has:
 
$$ \binom{9}{1} = 9 $$

different permutations. So, for each additional 2 an additional 0 needs to be included as well, which decreases the considered total length of the field resp. decreases the number of actions a stair-taker takes (0's are ignored). 

One main conclusion can be taken from this: 
 - For each combination out of (3.1) all permutations need to be evaluated and then summed up. For a single combination with a given number of 2's the permutations resp. combinations can be determined via:

$$ \binom{n - numberOf2s}{numberOf2s} $$

From 3.1 it is known that the number of 2's on *n* stairs ranges between 0,...,floor(*n* / 2), summing all combinations up results in:

$$(3.2)\hspace{10mm}NumOfCombinationsWithOrder(n) = \sum_{numberOf2s=0}^{floor(n / 2)} \binom{n - numberOf2s}{numberOf2s}$$


In [8]:
def numCombinationsWithOrderAndWithoutZeros(stairs):
    combis = 0       
    k = int(stairs / 2)
    for j in range(1, k+1):
        combis += scipy.special.binom(stairs - j, j)
    return int(combis + 1)

In [9]:
n = 10
print('Computing the number of combinations with order and without zeros for', n, 'stairs is', numCombinationsWithOrderAndWithoutZeros(n))

Computing the number of combinations with order and without zeros for 10 stairs is 89


For the example with 10 stairs the six unique combinations (3.1) result in the folling combinations with order according formula (3.2):

(Note: 0s are ignored)

$$\hspace{10mm}NumOfCombinationsWithOrder(10) = \sum_{numberOf2s=0}^{floor(10 / 2)} \binom{10 - numberOf2s}{numberOf2s}=$$ 

$$\binom{10 - 0}{0}+\binom{10 - 1}{1}+\binom{10 - 2}{2}+\binom{10 - 3}{3}+\binom{10 - 4}{4}+\binom{10 - 5}{5} = 1 + 9 + 28 + 35 + 15 + 1 = 89 $$

--- 

Breaking it down:

1. [1,1,1,1,1,1,1,1,1,1], *numberOf2s*=0:

$$\hspace{00mm}\sum_{numberOf2s=0}^{0} \binom{10 - 0}{0} = 1$$

---

2. [2,1,1,1,1,1,1,1,1,0], *numberOf2s*=1:
$$\hspace{0mm}[2,1,1,1,1,1,1,1,1,0]$$ 
$$\hspace{0mm}[1,2,1,1,1,1,1,1,1,0]$$
$$\hspace{0mm}...$$
$$\hspace{0mm}[1,1,1,1,1,1,1,1,2,0]$$

$$\hspace{00mm}\sum_{numberOf2s=1}^{1} \binom{10 - 1}{1} = \binom{9}{1} = 9$$

---

3. [2,2,1,1,1,1,1,1,0,0], *numberOf2s*=2:
$$\hspace{0mm}[2,2,1,1,1,1,1,1,0,0]$$ 
$$\hspace{0mm}[2,1,2,1,1,1,1,1,0,0]$$
$$\hspace{0mm}...$$
$$\hspace{0mm}[1,2,2,1,1,1,1,1,0,0]$$ 
$$\hspace{0mm}...$$
$$\hspace{0mm}[1,1,1,1,1,1,2,2,0,0]$$

$$\hspace{00mm}\sum_{numberOf2s=2}^{2} \binom{10 - 2}{2} = \binom{8}{2} = 28$$

---

4. [2,2,2,1,1,1,1,0,0,0], *numberOf2s*=3:
$$\hspace{0mm}[2,2,2,1,1,1,1,0,0,0]$$ 
$$\hspace{0mm}[2,2,1,2,1,1,1,0,0,0]$$
$$\hspace{0mm}...$$
$$\hspace{0mm}[1,2,2,2,1,1,1,0,0,0]$$ 
$$\hspace{0mm}...$$
$$\hspace{0mm}[1,1,1,1,2,2,2,0,0,0]$$

$$\hspace{00mm}\sum_{numberOf2s=3}^{3} \binom{10 - 3}{3} = \binom{7}{3} = 35$$

---

5. [2,2,2,2,1,1,0,0,0,0], *numberOf2s*=4:
$$\hspace{0mm}[2,2,2,2,1,1,0,0,0,0]$$ 
$$\hspace{0mm}[2,2,2,1,2,1,0,0,0,0]$$
$$\hspace{0mm}...$$
$$\hspace{0mm}[1,2,2,2,2,1,0,0,0,0]$$ 
$$\hspace{0mm}...$$
$$\hspace{0mm}[1,1,2,2,2,2,0,0,0,0]$$

$$\hspace{00mm}\sum_{numberOf2s=4}^{4} \binom{10 - 4}{4} = \binom{6}{4} = 15$$

---

5. [2,2,2,2,2,0,0,0,0,0], *numberOf2s*=5:
$$\hspace{0mm}[2,2,2,2,2,0,0,0,0,0]$$ 

$$\hspace{00mm}\sum_{numberOf2s=5}^{5} \binom{10 - 5}{5} = \binom{5}{5} = 1$$


# Comparing results with the fibonacci-solution
Part 4.

In [11]:
def numCombinationsAccFibonacci(n):
    if n <= 1:
        return 1
    return numCombinationsAccFibonacci(n-1) + numCombinationsAccFibonacci(n-2)

In [49]:
myRange = [1, 5, 10, 15, 20]
df = pd.DataFrame()
df['numOfStairs'] = [(i) for i in myRange]  
df['numCombisAccFibonacci'] = [numCombinationsAccFibonacci(i) for i in myRange]  
df['numCombisWithOrderWithoutZeros'] = [numCombinationsWithOrderAndWithoutZeros(i) for i in myRange] 
df

numOfStairs  numCombisAccFibonacci  numCombisWithOrderWithoutZeros
0            1                      1                               1
1            5                      8                               8
2           10                     89                              89
3           15                    987                             987
4           20                  10946                           10946

# Considering a pause option when having number of stairs' actions of taking 1, 2 or 0 steps
Part 5. Combinations of the **third level**. The order, where 2 steps insted of 1 and where a pause resp. zeros is consumed relevent.

For each time the stair-taker takes two steps instead of one also a pause(a 0) needs to be added to the field. Where the pause is taken should also now be considered. For each unique combination with given elements (3.1) two orderings need to be considered:
 - second level: having the 2's in all possible positions for each combination of (3.1)
 - third level:  having the 0's in all possible positions for each combination of (3.2)
 
Considering the second combination with one 2 led to 9 different combinations with order:
$$\hspace{0mm}[2,1,1,1,1,1,1,1,1,0]$$ 
$$\hspace{0mm}[1,2,1,1,1,1,1,1,1,0]$$
$$\hspace{0mm}...$$
$$\hspace{0mm}[1,1,1,1,1,1,1,2,1,0]$$
$$\hspace{0mm}[1,1,1,1,1,1,1,1,2,0]$$

The number of combinations having 2's on all positions is (ignoring the 0):

$$ \binom{9}{1} = 9 $$

For the third level, for each of these combinations again all possible permutations of 0's need to be computed. The number of 0's and 2' is always equal. It follows that the zero can permutate on all position in the field, which is:

$$ \binom{10}{1} = 10 $$

For example, taking the first combination leads to the following 10 combinations with permutating the 0:

$$\hspace{0mm}[2,1,1,1,1,1,1,1,1,0]$$
$$\hspace{0mm}[2,1,1,1,1,1,1,1,0,1]$$ 
$$\hspace{0mm}....$$
$$\hspace{0mm}[2,1,0,1,1,1,1,1,1,1]$$
$$\hspace{0mm}[2,0,1,1,1,1,1,1,1,1]$$
$$\hspace{0mm}[0,2,1,1,1,1,1,1,1,1]$$

One main conclusion can be taken from this: 
 - For each combination out of (3.2) all permutations (permutating 0's) need to be evaluated.
 
Due to the fact that the number of 0's and 2' is equal it follows:

$$(5.1)\hspace{10mm}NumOfCombinationsWithOrder(n) = \sum_{numberOf2s=0}^{floor(n / 2)} \binom{n - numberOf2s}{numberOf2s} * \binom{n}{numberOf2s}$$

In [15]:
def numCombinationsWithOrderAndWithZeros(stairs):
    _combis = 0      
    _k = int(stairs / 2)
    for j in range(1, _k+1):
        _combis += scipy.special.binom(stairs - j*(2-1), j) * scipy.special.binom(stairs, j)
    return int(_combis + 1)

In [9]:
numCombinationsWithOrderAndWithZeros(10)

8953

For the example with 10 stairs the six unique combinations (3.1) result in 89 combinations with order and not considering the zero (3.2). Considering the zero leads to the folling combinations with order according formula (5.1):

$$\hspace{10mm}NumOfCombinationsWithOrder(10) = \sum_{numberOf2s=0}^{floor(10 / 2)} \binom{10 - numberOf2s}{numberOf2s} * \binom{10}{numberOf2s} =$$

$$\binom{10 - 0}{0}*\binom{10 - 0}{0} + \binom{10 - 1}{1}*\binom{10}{1} + \binom{10 - 2}{2}*\binom{10}{2} + \binom{10 - 3}{3}*\binom{10}{3} + \binom{10 - 4}{4}*\binom{10}{4} + \binom{10 - 5}{5}*\binom{10}{5} = $$

$$1*1 + 9*10 + 28*45 + 35*120 + 15*210 + 1*252 = 8953 $$

--- 

Breaking it down:

1. [1,1,1,1,1,1,1,1,1,1], *numberOf2s*=0:

$$\hspace{00mm}\sum_{numberOf2s=0}^{0} \binom{10 - 0}{0} * \binom{10}{0} = 1$$

---

2. [2,1,1,1,1,1,1,1,1,0], *numberOf2s*=1:
$$\hspace{00mm}[2,1,1,1,1,1,1,1,1,0]$$
$$\hspace{20mm}[2,1,1,1,1,1,1,1,0,1]$$
$$\hspace{20mm}[2,1,1,1,1,1,1,0,1,1]$$
$$\hspace{20mm}....$$
$$\hspace{20mm}[0,2,1,1,1,1,1,1,1,1]$$
$$\hspace{00mm}[1,2,1,1,1,1,1,1,1,0]$$
$$\hspace{20mm}[1,2,1,1,1,1,1,1,0,1]$$
$$\hspace{20mm}....$$
$$\hspace{20mm}[0,2,1,1,1,1,1,1,1,1]$$

$$\hspace{00mm}\sum_{numberOf2s=1}^{1} \binom{10 - 1}{1} * \binom{10}{1}  = \binom{9}{1} * \binom{10}{1} = 90$$

---

3. [2,2,1,1,1,1,1,1,0,0], *numberOf2s*=2:
$$\hspace{00mm}[2,2,1,1,1,1,1,1,0,0]$$
$$\hspace{20mm}[2,2,1,1,1,1,1,0,1,0]$$
$$\hspace{20mm}[2,2,1,1,1,1,0,1,1,0]$$
$$\hspace{20mm}....$$
$$\hspace{00mm}[2,1,2,1,1,1,1,1,0,0]$$
$$\hspace{20mm}[2,1,2,1,1,1,1,0,1,0]$$ 
$$\hspace{20mm}...$$

$$\hspace{00mm}\sum_{numberOf2s=1}^{1} \binom{10 - 2}{2} * \binom{10}{2}  = \binom{8}{2} * \binom{10}{2} = 1260$$

---

4. [2,2,2,1,1,1,1,0,0,0], *numberOf2s*=3:
$$\hspace{00mm}[2,2,2,1,1,1,1,0,0,0]$$ 
$$\hspace{20mm}[2,2,2,1,1,1,0,1,0,0]$$
$$\hspace{20mm}[2,2,2,1,1,0,1,1,0,0]$$
$$\hspace{20mm}...$$
$$\hspace{00mm}[1,2,2,2,1,1,1,0,0,0]$$ 
$$\hspace{20mm}[1,2,2,2,1,1,0,1,0,0]$$
$$\hspace{20mm}...$$

$$\hspace{00mm}\sum_{numberOf2s=3}^{3} \binom{10 - 3}{3} * \binom{10}{3}  = \binom{7}{3} * \binom{10}{3} = 4200$$

---

5. [2,2,2,2,1,1,0,0,0,0], *numberOf2s*=4:
$$\hspace{00mm}[2,2,2,2,1,1,0,0,0,0]$$ 
$$\hspace{20mm}[2,2,2,2,1,0,1,0,0,0]$$ 
$$\hspace{20mm}[2,2,2,2,0,1,0,0,0,0]$$ 
$$\hspace{20mm}...$$ 

$$\hspace{00mm}\sum_{numberOf2s=4}^{4} \binom{10 - 4}{4} * \binom{10}{4}  = \binom{6}{4} * \binom{10}{4} = 3150$$

---

5. [2,2,2,2,2,0,0,0,0,0], *numberOf2s*=5:
$$\hspace{00mm}[2,2,2,2,2,0,0,0,0,0]$$ 
$$\hspace{20mm}[2,2,2,2,0,2,0,0,0,0]$$ 
$$\hspace{20mm}[2,2,2,0,2,2,0,0,0,0]$$ 
$$\hspace{20mm}...$$ 

$$\hspace{00mm}\sum_{numberOf2s=5}^{5} \binom{10 - 5}{5} * \binom{10}{5} = \binom{5}{5} * \binom{10}{5} = 252$$


In [51]:
class StairCaseCycleEvaluation: 
    def __init__(self):
        self.__field = None
    
    def NextCombination(self, events, level, outcomes):
        self.__field[level] = (self.__field[level] + 1) % outcomes
        if self.__field[level] == 0 and level + 1 < events:
            self.NextCombination(events, level + 1, outcomes)
 
    def ComputeNumOfCombinations(self, numOfStairsteps, numMaxSteps):
        self.__field = [0 for x in range(numOfStairsteps)] 
        totals = numMaxSteps**numOfStairsteps
        numberOfHits = 0
        combi = 0
                
        if totals <= 1000000:
            while(True):
                numberOfHits += 1 if sum(self.__field) == numOfStairsteps else 0
                self.NextCombination(numOfStairsteps, 0, numMaxSteps)
                combi += 1
                
                if combi >= totals:
                    break
            return int(numberOfHits)
        else:
            return None

Cycle-evaluation functionality for the Stair-Case-Problem. Note: The time complexity is O(3^*n*).

In [52]:
myStairCaseProblem = StairCaseCycleEvaluation()

df['numCombisWithOrderWithZeros'] = [numCombinationsWithOrderAndWithZeros(i) for i in myRange] 
df['numCombisWithOrderWithZerosCycleEval'] = [myStairCaseProblem.ComputeNumOfCombinations(i,3) for i in myRange] 
df

numOfStairs  numCombisAccFibonacci  numCombisWithOrderWithoutZeros  \
0            1                      1                               1   
1            5                      8                               8   
2           10                     89                              89   
3           15                    987                             987   
4           20                  10946                           10946   

   numCombisWithOrderWithZeros  numCombisWithOrderWithZerosCycleEval  
0                            1                                   1.0  
1                           51                                  51.0  
2                         8953                                8953.0  
3                      1787607                                   NaN  
4                    377379369                                   NaN

The cycle-evaluation was only applied up to a total cyclye-length of 1 million due to the high time complexity. Here, either [**pypy**](https://towardsdatascience.com/run-your-python-code-as-fast-as-c-4ae49935a826) or [**executing the code in C++/C#**](https://nbviewer.jupyter.org/github/Gordi33/The-Laws-of-the-Game/blob/master/DistributionComputation.ipynb) is recommended.
The cycle-evaluation is used to verify the results numerically.